compiler/rustc_resolve/src/late.rs

```diff
@@ -1787,6 +1805,32 @@ impl<'a: 'ast, 'b, 'ast> LateResolutionVisitor<'a, 'b, 'ast> {
         }
     }
 
+    /// For backwards compatibility, we need to accept the following:
+    ///
+    /// ```
+    /// trait AsStr {
+    ///     fn as_str(&self) -> &str;
+    /// }
+    ///
+    /// fn foo(a: &(dyn AsStr + '_)) -> &str { a.as_str() }
+    /// ```
+    ///
+    /// When we're resolving function parameters, we usually want each elided or
+    /// anonymous lifetime to result in a new elision candidate, but not in this case.
+    #[instrument(level = "trace", skip(self))]
+    fn resolve_anonymous_lifetime_in_trait_object_bounds(&mut self, lifetime: &Lifetime) {
+        debug_assert_eq!(lifetime.ident.name, kw::UnderscoreLifetime);
+        debug_assert!(self.diagnostic_metadata.current_trait_object.is_some());
+
+        if self.in_func_params {
+            self.with_lifetime_rib(LifetimeRibKind::Elided(LifetimeRes::Infer), |this| {
```

Comment:
The resolution is not `Infer`. In HIR, we should get an `ImplicitObjectLifetimeDefault`, but I did not create such variant in `LifetimeRes`. Could you create an extra `ImplicitObjectLifetimeDefault` variant in `LifetimeRes` and adapt lowering to translate it to `LifetimeName::ImplicitObjectLifetimeDefault`?

`Infer` means that type-checking will try to find the correct lifetime through type-inference. This is typically what appears in function bodies. `ImplicitObjectLifetimeDefault` means that we will inspect the surrounding type to apply the object lifetime rules: `&'a (dyn Trait + '_)` implicitly becomes `&'a (dyn Trait + 'a)` etc.

Reply:
Thank you for the pointer!

This does indeed get us past any errors in resolve and astconv, and all the tests I wrote compile as I had intended.

However, it seems `ImplicitObjectLifetimeDefault` results in the wrong lifetime for `Box<dyn Trait + '_>`, as this failed test shows:

```rust
// check-pass

struct Ctxt {
    v: usize,
}

trait GetCtxt {
    // Here the `&` is bound in the method definition:
    fn get_ctxt(&self) -> &Ctxt;
}

struct HasCtxt<'a> {
    c: &'a Ctxt,
}

impl<'a> GetCtxt for HasCtxt<'a> {
    // Ok: Have implied bound of WF(&'b HasCtxt<'a>)
    // so know 'a: 'b
    // so know &'a Ctxt <: &'b Ctxt
    fn get_ctxt<'b>(&'b self) -> &'a Ctxt {
        self.c
    }
}

fn get_v(gc: Box<dyn GetCtxt + '_>) -> usize {
    gc.get_ctxt().v
}

fn main() {
    let ctxt = Ctxt { v: 22 };
    let hc = HasCtxt { c: &ctxt };
    assert_eq!(get_v(Box::new(hc) as Box<dyn GetCtxt>), 22);
}
```

```
---- [ui] src/test/ui/regions/regions-trait-1.rs stdout ----

error: test compilation failed although it shouldn't!
status: exit status: 1
command: "/Users/ben/Documents/Rust/rust-lang/rust/build/aarch64-apple-darwin/stage1/bin/rustc" "/Users/ben/Documents/Rust/rust-lang/rust/src/test/ui/regions/regions-trait-1.rs" "-Zthreads=1" "--target=aarch64-apple-darwin" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/Users/ben/Documents/Rust/rust-lang/rust/build/aarch64-apple-darwin/test/ui/regions/regions-trait-1" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/Users/ben/Documents/Rust/rust-lang/rust/build/aarch64-apple-darwin/native/rust-test-helpers" "-L" "/Users/ben/Documents/Rust/rust-lang/rust/build/aarch64-apple-darwin/test/ui/regions/regions-trait-1/auxiliary"
stdout: none
--- stderr -------------------------------
error[E0597]: `ctxt` does not live long enough
  --> /Users/ben/Documents/Rust/rust-lang/rust/src/test/ui/regions/regions-trait-1.rs:31:27
   |
LL |     let hc = HasCtxt { c: &ctxt };
   |                           ^^^^^ borrowed value does not live long enough
LL |     assert_eq!(get_v(Box::new(hc) as Box<dyn GetCtxt>), 22);
   |                      ------------ cast requires that `ctxt` is borrowed for `'static`
LL | }
   | - `ctxt` dropped here while still borrowed

error: aborting due to previous error

For more information about this error, try `rustc --explain E0597`.
------------------------------------------
```

So, I'm not sure where to go from there.

Reply:
^^ This is the more concerning bit. Please advise when you can.